### PR TITLE
bpo-44026: Idle get Python's 'did you mean'

### PR DESCRIPTION
**Pull request by E-Paine**
Based on the test_exception code included on bpo by @sweeneyde. If the exception is an `AttributeError` or `NameError`, we remove the last line from the normal traceback and replace it with one captured from `sys.__excepthook__`. This implementation utilises `test.support.captured_stderr`, like the original.

https://bugs.python.org/issue44026

### CODE REVIEW

**Comment by E-Paine** [comment]
Pipelines failure is an unrelated `test_concurrent_futures` issue

**Comment by shreyanavigyan** [comment]
AFAICT this PR needs a news entry

**Comment by E-Paine** [comment]
The pipelines test is a different `test_concurrent_futures` failure. I cannot see how these failures could be linked to this PR so am inclined to ignore them.

**Comment by sweeneyde** [comment]
Perhaps it would be better to use `contextlib.redirect_stderr(to)`, since, from the docs:

```
Note test.support is not a public module. It is documented here to help Python developers write tests. The API of this module is subject to change without backwards compatibility concerns between releases.
```

**Comment by E-Paine** [comment]
> Perhaps it would be better to use `contextlib.redirect_stderr(to)`

I did not know that existed; thanks for pointing me to it!

**Comment by E-Paine** [comment]
OK, that's the third consecutive pipelines Windows failure of `test_concurrent_futures`. Is this an issue with other PRs, or could this (somehow) be to do with my changes? I'll try a merge and see if that fixes it.

**Comment by shreyanavigyan** [comment]
The test you're talking about has failed in lot of PRs recently.

**Comment by terryjreedy** [comment]
Paine, I am currently reviewing and revising, so please no more changes.

**Review by terryjreedy — APPROVED** [review]
Functions should only be nested if they need to be.  They cannot be tested properly if they are.   I expanded the scope slightly to make it more 'complete'.  I then reduced the scope of the new tests to testing the new unit, and used subTest to combine and add a 'normal' case.

print_exc should be checked to see if *it* should be nested.  get_message_lines should become get_message.  Currently, format_exception_only only returns multiple lines.  Combining them could be done within the function.  I left that for another issue.  

I think this is ready to merge, but am leaving for others and myself to look at again (and, I hope, for someone else to fix the test failures if not done yet).

Current coverage is 54% on my machine.  Should update if any other edits.

**Comment by E-Paine** [comment]
Sorry Terry, I know you asked for no more changes, but I need to patch it to correctly deal with multiple errors in a traceback (I will also add a test for this).

```
>>> try:
...     abc
... except NameError:
...     bcd
... 
...     
Traceback (most recent call last):
  File "<pyshell#8>", line 2, in <module>
    abc
NameError: name 'bcd' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<pyshell#8>", line 4, in <module>
    bcd
NameError: name 'bcd' is not defined
```

**Comment by E-Paine** [comment]
MacOS failure is an unrelated issue with `test_ssl`.

**Comment by terryjreedy** [comment]
Sorry, I meant no more changes *until I posted my edits*, to avoid any merge conflicts.  Once I did, I did not merge immediately to allow for further review and possible change.  Good thing.

In your bad example above, should not the first bad name have been 'abc'?  Was 'bcd' a result of fetching sys.exec info again, in get_message_lines, during processing?  In any case, with the last patch:
```
try: abc
except NameError: ing

Traceback (most recent call last):
  File "<pyshell#2>", line 1, in <module>
    try: abc
NameError: name 'abc' is not defined. Did you mean: 'abs'?

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<pyshell#2>", line 2, in <module>
    except NameError: ing
NameError: name 'ing' is not defined. Did you mean: 'int'?
```
Both bad names are listed and corrected.  Great.

I am impressed by and like the new test, but want to try to improve it and the previous one further.  So no changes while I do ;-).  (I am assuming that you are not editing right now.)

File: Lib/idlelib/idle_test/test_run.py [hunk 1]
```diff
@@ -51,14 +51,47 @@ def test_get_message(self):
...
+                ('0/1', 'int.reel', ZeroDivisionError, AttributeError,
+                 "division by zero\n",
+                 "type object 'int' has no attribute 'reel'. Did you mean: 'real'?\n"),
+                ('abc', 'int.reel', NameError, AttributeError,
+                 "name 'abc' is not defined. Did you mean: 'abs'?\n",
+                 "type object 'int' has no attribute 'reel'. Did you mean: 'real'?\n"),
+                )
+        for code1, code2, exc1, exc2, msg1, msg2 in data:
+            with self.subTest(code=code1):
+                with self.subTest(code=code2):
```
Comment (terryjreedy) [inline Lib/idlelib/idle_test/test_run.py]: The arguments to subTest are only labels for failure reports.  'code=code2' masks 'code=code1'.  I replaced above with one `with self.subTest(codes=(code1,code2):` so any failure is labelled with the pair.  (And dedented body.)

File: Lib/idlelib/idle_test/test_run.py [hunk 2]
```diff
@@ -51,14 +51,47 @@ def test_get_message(self):
...
                     self.assertEqual(actual, expect)
                 else:
                     unittest.skip("Polluted namespace")
                     # Should make impossible.
 
+    def test_get_multiple_message(self):
+        data = (('abc', '1/0', NameError, ZeroDivisionError,
+                 "name 'abc' is not defined. Did you mean: 'abs'?\n",
+                 "division by zero\n"),
+                ('0/1', 'int.reel', ZeroDivisionError, AttributeError,
```
Comment (terryjreedy) [inline Lib/idlelib/idle_test/test_run.py]: '0/1' does not raise ;-).  Test have to be tested that they actually test ;-).  I found this after putting `print(code1, code2, flush=True)` at the bottom of the subtest.

File: Lib/idlelib/idle_test/test_run.py [hunk 3]
```diff
@@ -51,14 +51,47 @@ def test_get_message(self):
...
+                        except exc1:
+                            try:
+                                eval(compile(code2, '', 'eval'))
+                            except exc2:
+                                with captured_stderr() as output:
+                                    run.print_exception()
+                                actual = output.getvalue()
+                                self.assertIn(msg1, actual)
+                                self.assertIn(msg2, actual)
+                                return
```
Comment (terryjreedy) [inline Lib/idlelib/idle_test/test_run.py]: This exits after 1 subtest, as discovered by print mentioned above.

File: Lib/idlelib/idle_test/test_run.py [hunk 4]
```diff
@@ -51,14 +51,47 @@ def test_get_message(self):
...
+        unittest.skip("Polluted namespace")
```
Comment (terryjreedy) [inline Lib/idlelib/idle_test/test_run.py]: I am removing this.  We are passing the function local namespace to eval.

**Review by terryjreedy — APPROVED** [review]
I did edits in comprehensible chunks, testing after each. I will again wait a bit for you to check before merging.

**Comment by E-Paine** [comment]
> [from the news] Include parser's typo fix suggestions

Just checking: is it the parser?

**Comment by terryjreedy** [comment]
@E-Paine The conflict was in the idlelib NEWS file.  I fixed it and then remembered that this was not be be backported to 3.9 because it does not have the hints.

**Comment by terryjreedy** [comment]
Buildbot failure due to test asyncio deleted.

**Pull request merged**